File: ios/Classes/ARCameraPoseStreamHandler.swift
//
//  ARCameraPoseStreamHandler.swift
//  ar_flutter_plugin
//
//  Created by Nick Kurochkin on 16.09.2022.
//

import Foundation
import Combine
import ARKit
import Flutter

class ARCameraPoseStreamHandler: NSObject,  FlutterStreamHandler {
    var sink: FlutterEventSink?
    
    var lastFrame: ARFrame?
    var lastVisibleNodes: [SCNNode]?
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
    
    func updateCameraPose(frame: ARFrame) {
        guard let sink = sink else {
            return
        }
        sink(serializeARCamera(frame.camera))
    }
}

File: android/src/main/kotlin/io/carius/lars/ar_flutter_plugin/ARCameraPoseStreamHandler.kt
package io.carius.lars.ar_flutter_plugin
import com.google.ar.core.Pose
import io.carius.lars.ar_flutter_plugin.Serialization.serializeCameraPoseInfo
import io.flutter.plugin.common.EventChannel

class ARCameraPoseStreamHandler: EventChannel.StreamHandler {
    var sink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    fun updateCameraPose(pose: Pose) {
        sink?.success(serializeCameraPoseInfo(pose))
    }
}